compiler/src/iree/compiler/Dialect/HAL/Transforms/ResolveTopologyQueries.cpp

```diff
@@ -0,0 +1,228 @@
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-hal-resolve-topology-queries"
+
+static llvm::cl::opt<bool> clExternalResourcesMappable(
+    "iree-stream-external-resources-mappable",
+    llvm::cl::desc("Allocates external resources as host-visible and mappable. "
+                   "This can degrade performance and introduce allocation "
+                   "overhead and staging buffers for readback on the host "
+                   "should be managed by the calling application instead."),
+    llvm::cl::init(false));
+
+namespace mlir::iree_compiler::IREE::HAL {
+
+#define GEN_PASS_DEF_RESOLVETOPOLOGYQUERIESPASS
+#include "iree/compiler/Dialect/HAL/Transforms/Passes.h.inc"
+
+namespace {
+
+// Maps a resource type to the corresponding HAL memory types and buffer usage.
+// This will fail if the resource type is not directly mappable to HAL bits.
+// The bits set here represent the superset of required and allowed bits and
+// are useful for providing buffers back to users via the ABI that may need to
+// be used for more than just what the internal program requires.
+LogicalResult
+deriveAllowedResourceBufferBits(Location loc, IREE::HAL::Lifetime lifetime,
+                                IREE::HAL::MemoryTypeBitfield &memoryTypes,
+                                IREE::HAL::BufferUsageBitfield &bufferUsage) {
+  memoryTypes = IREE::HAL::MemoryTypeBitfield::None;
+  bufferUsage = IREE::HAL::BufferUsageBitfield::None;
+  if (failed(deriveRequiredResourceBufferBits(loc, lifetime, memoryTypes,
+                                              bufferUsage))) {
+    return failure();
+  }
+  switch (lifetime) {
+  default:
+    break;
+  case IREE::HAL::Lifetime::External:
+    if (clExternalResourcesMappable) {
+      // #yolo; these come from/go to outside the program.
+      // Today we assume they are device-local|host-visible just for
+      // practical purposes but that does not have to be true. We really
+      // want this to be something we analyze and handle on the edges
+      // (transferring devices/etc if needed).
+      memoryTypes = memoryTypes | IREE::HAL::MemoryTypeBitfield::DeviceLocal |
+                    IREE::HAL::MemoryTypeBitfield::HostVisible;
+      // NOTE: we may not map it but users may after they get them back.
+      // Another reason we should annotate this - having a buffer be
+      // mappable is potentially expensive (may get a 2nd copy in memory!).
+      bufferUsage = bufferUsage | IREE::HAL::BufferUsageBitfield::Mapping;
+    } else {
+      memoryTypes = memoryTypes | IREE::HAL::MemoryTypeBitfield::DeviceLocal;
+    }
+    break;
+  }
+  return success();
+}
+
+// Checks if all affinities of the optimal attribute refer to the same device
```

Comment:
nit: "check" can be ambiguous and what a reader is mostly interested in with functions like this is what they return.
```suggestion
// Returns true if all affinities of the optimal attribute refer to the same device
```
(here and below)